### PR TITLE
dial: add `DialContext` function

### PR DESCRIPTION
**Pull request by DerekBum**
In order to replace timeouts with contexts in `Connect` instance creation (go-tarantool), I need a `DialContext` function. It accepts context, and cancels, if context is canceled by user.

Disabled flaking CI Windows runner. Because we don't support connector on Windows, I considered disabling these tests.

Part of https://github.com/tarantool/go-tarantool/issues/136

### CODE REVIEW

File: net.go [hunk 1]
```diff
@@ -109,11 +122,11 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
...
 	session []byte) (*Conn, error) {
 	var d net.Dialer
-	return dialSession(d, network, addr, ctx, flags, session)
+	return dialSession(d, context.Background(), network, addr, ctx, flags, session)
 }
 
-func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
-	session []byte) (*Conn, error) {
+func dialSession(d net.Dialer, context context.Context, network, addr string,
+	ctx *Ctx, flags DialFlags, session []byte) (*Conn, error) {
```
Comment (oleg-jukovec) [inline net.go]: Let's stop adding unnecessary arguments here and reorganize the function a little.

For example, it could be something like this:

1.
```suggestion
func createSession(c net.Conn, ctx *Ctx, flags DialFlags, session []byte) (*Conn, error) {
```

2. A separate function for `ctx *Ctx` preparation:
https://github.com/tarantool/go-openssl/blob/bdadb815fbb91fdee2b475f7f38b17a7f56db0e0/net.go#L134-L141

3. A separate function for `c net.Conn` creation:
https://github.com/tarantool/go-openssl/blob/bdadb815fbb91fdee2b475f7f38b17a7f56db0e0/net.go#L130-L133
and
https://github.com/tarantool/go-openssl/blob/bdadb815fbb91fdee2b475f7f38b17a7f56db0e0/net.go#L143-L146

or even
```suggestion
func setSession(client *Conn, flags DialFlags, session []byte) (*Conn, error) {
```

Comment (oleg-jukovec) [reply net.go]: Or you can split this function in some other way.

File: net.go [hunk 2]
```diff
@@ -90,7 +91,19 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
...
+	return dialSession(d, context.Background(), network, addr, ctx, flags, nil)
+}
+
+// DialContext acts like Dial but takes a context for network dial.
+//
+// The context includes only network dial. It does not include OpenSSL calls.
+//
+// See func Dial for a description of the network, addr, ctx and flags
+// parameters.
+func DialContext(context context.Context, network, addr string,
```
Comment (oleg-jukovec) [inline net.go]: It would be nice to see a simple test for the call if it's not too complicated.

File: net.go (hunk 1 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go (hunk 2 above)
Comment (DerekBum) [reply net.go]: New tests added into `net_test.go`.

File: net.go [hunk 3]
```diff
@@ -90,7 +91,33 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
 func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
 	flags DialFlags) (*Conn, error) {
 	d := net.Dialer{Timeout: timeout}
-	return dialSession(d, network, addr, ctx, flags, nil)
+	if err := prepareCtx(ctx); err != nil {
```
Comment (oleg-jukovec) [inline net.go]: The value `ctx` will not be updated.
```suggestion
	if ctx, err := prepareCtx(ctx); err != nil {
```

File: net.go [hunk 4]
```diff
@@ -109,58 +136,81 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
...
+			return err
 		}
 		// TODO: use operating system default certificate chain?
 	}
+	return nil
+}
 
-	c, err := d.Dial(network, addr)
+func createConnection(d net.Dialer, context context.Context, network,
+	addr string) (net.Conn, error) {
```
Comment (oleg-jukovec) [inline net.go]: `context.Context` should be always first arg.

File: net.go [hunk 5]
```diff
@@ -109,58 +136,81 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
...
-	}
+	return c, nil
+}
+
+func applyFlags(conn *Conn, host string, flags DialFlags) error {
+	var err error
 	if flags&DisableSNI == 0 {
 		err = conn.SetTlsExtHostName(host)
 		if err != nil {
 			conn.Close()
```
Comment (oleg-jukovec) [inline net.go]: We open the connection out of the scope of the function. So let and close it out of the scope in the error case.

This will make it easier to understand what is happening in the error case from the caller code.

File: net.go [hunk 6]
```diff
@@ -109,58 +136,81 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
...
+
+func createSession(c net.Conn, flags DialFlags, addr string, ctx *Ctx,
+	session []byte) (*Conn, error) {
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := Client(c, ctx)
+	if err != nil {
+		c.Close()
```
Comment (oleg-jukovec) [inline net.go]: The same about `c` and the `createSession`.

File: net.go (hunk 3 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go (hunk 4 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go (hunk 5 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go (hunk 6 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go [hunk 7]
```diff
@@ -90,7 +91,37 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
...
-	return dialSession(d, network, addr, ctx, flags, nil)
+	ctx, err := prepareCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := createConnection(context.Background(), d, network, addr)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
```
Comment (oleg-jukovec) [inline net.go]: Nit, I don't sure the close of the network connection at the end of the call is a good idea. It should not work.

File: net.go [hunk 8]
```diff
@@ -90,7 +91,37 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
...
+	return createSession(conn, flags, addr, ctx, nil)
```
Comment (oleg-jukovec) [inline net.go]:
```suggestion
	c, err := createSession(conn, flags, addr, ctx, nil)
	if err != nil {
	    conn.Close()
	    return nil, err
	}
	return c, nil
```

File: net.go [hunk 9]
```diff
@@ -126,41 +161,58 @@ func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
...
+	session []byte) (*Conn, error) {
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := Client(c, ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
```
Comment (oleg-jukovec) [inline net.go]: It closes the connection at the end of the function. But we need an open connection.

Please, update tests to handle the situation.

File: net.go (hunk 7 above)
Comment (DerekBum) [reply net.go]: Thanks, updated.

File: net.go (hunk 8 above)
Comment (DerekBum) [reply net.go]: Thanks, updated.

File: net.go (hunk 9 above)
Comment (DerekBum) [reply net.go]: New test added.

File: net.go [hunk 10]
```diff
@@ -126,41 +170,59 @@ func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
...
-		err := conn.setSession(session)
-		if err != nil {
-			c.Close()
-			return nil, err
-		}
-	}
+	return c, nil
+}
+
+func applyFlags(conn *Conn, host string, flags DialFlags) error {
```
Comment (oleg-jukovec) [inline net.go]:
```suggestion
func handshake(conn *Conn, host string, flags DialFlags) error {
```

The essence of the function not about flags applying.

File: net.go [hunk 11]
```diff
@@ -126,41 +170,59 @@ func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
...
+	}
+	return nil
+}
+
+func createSession(c net.Conn, flags DialFlags, addr string, ctx *Ctx,
+	session []byte) (*Conn, error) {
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
```
Comment (oleg-jukovec) [inline net.go]: The code is about a network connection/possibility of a network connection. Not about the create TLS session from the connection.

It is better to move it into `createConnection()` or a separate call.

```suggestion
func createSession(c net.Conn, flags DialFlags, host string, ctx *Ctx,
	session []byte) (*Conn, error) {
```

File: net_test.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+		}()
+	}
+}
+
+func TestDial(t *testing.T) {
+	ctx := getCtx(t)
+	if err := ctx.SetCipherList("AES128-SHA"); err != nil {
+		t.Fatal(err)
+	}
+	ssl_listener, err := Listen("tcp", "localhost:0", ctx)
```
Comment (oleg-jukovec) [inline net_test.go]: See linter errors.

File: net_test.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+	go sslConnect(t, ssl_listener)
```
Comment (oleg-jukovec) [inline net_test.go]: It would be nice to wait for the end of the gorouitine execution (with WaitGroup or a channel).

File: net.go [hunk 14]
```diff
@@ -126,41 +170,59 @@ func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
 		}
 		// TODO: use operating system default certificate chain?
 	}
+	return ctx, nil
+}
 
-	c, err := d.Dial(network, addr)
+func createConnection(context context.Context, d net.Dialer, network,
+	addr string) (net.Conn, error) {
+	c, err := d.DialContext(context, network, addr)
```
Comment (oleg-jukovec) [inline net.go]: It could be simplified and unified in such way:
```suggestion
func createConnection(context context.Context, network, addr string) (net.Conn, error) {
	return net.Dialer{}.DialContext(context, network, addr)
```
and usage in the timeout case (`DialTimeout`) will be:
```
ctx, cancel := context.WithTimeout(context.Background(), timeout)
defer cancel()
conn, err := createConnection(ctx, network, addr)
```

Or you could even don't create a separate function and just use different `net.*` calls for different cases.
`Dial()/DialTimeout()/DialContext()` -> `net.Dial()/net.DialTimeout()/net.Dialer{}.DialContext()`.

File: net.go [hunk 15]
```diff
@@ -90,7 +91,43 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
...
+	ctx *Ctx, flags DialFlags) (*Conn, error) {
+	d := net.Dialer{}
+	ctx, err := prepareCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := createConnection(context, d, network, addr)
+	if err != nil {
+		return nil, err
+	}
```
Comment (oleg-jukovec) [inline net.go]: It is very likely that a connection will fail, so we will not need a ssl context object.
```suggestion
	conn, err := createConnection(context, d, network, addr)
	if err != nil {
		return nil, err
	}
	ctx, err := prepareCtx(ctx)
	if err != nil {
		return nil, err
	}
```

Comment (DerekBum) [reply net.go]: Updated.

File: net.go (hunk 14 above)
Comment (DerekBum) [reply net.go]: Updated. Now I use `net.*` calls and host parsing is in a separate function now.

File: net_test.go (hunk 13 above)
Comment (DerekBum) [reply net_test.go]: Updated.

File: net_test.go (hunk 12 above)
Comment (DerekBum) [reply net_test.go]: Fixed.

File: net.go (hunk 11 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go (hunk 10 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go [hunk 16]
```diff
@@ -89,8 +90,53 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
...
+
+	conn, err := net.DialTimeout(network, addr, timeout)
+	if err != nil {
+		return nil, err
+	}
+	ctx, err = prepareCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
```
Comment (oleg-jukovec) [inline net.go]:
```suggestion
	if err != nil {
		conn.Close()
		return nil, err
	}
```

File: net.go [hunk 17]
```diff
@@ -89,8 +90,53 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
...
+
+	dialer := net.Dialer{}
+	conn, err := dialer.DialContext(context, network, addr)
+	if err != nil {
+		return nil, err
+	}
+	ctx, err = prepareCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
```
Comment (oleg-jukovec) [inline net.go]:
```suggestion
	if err != nil {
		conn.Close()
		return nil, err
	}
```

File: net.go [hunk 18]
```diff
@@ -108,59 +154,80 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
...
-		// TODO: use operating system default certificate chain?
-	}
 
-	c, err := d.Dial(network, addr)
+	conn, err := net.Dial(network, addr)
 	if err != nil {
 		return nil, err
 	}
-	conn, err := Client(c, ctx)
+	ctx, err = prepareCtx(ctx)
```
Comment (oleg-jukovec) [inline net.go]: +conn.Close() as the same.

File: net.go [hunk 19]
```diff
@@ -108,59 +154,80 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
...
+	if ctx == nil {
+		var err error
+		ctx, err = NewCtx()
 		if err != nil {
-			c.Close()
 			return nil, err
 		}
+		// TODO: use operating system default certificate chain?
 	}
+	return ctx, nil
```
Comment (oleg-jukovec) [inline net.go]:
```suggestion
if ctx == nil {
    return NewCtx()
}
return ctx, nil
```

File: net_test.go [hunk 20]
```diff
@@ -0,0 +1,94 @@
+package openssl
```
Comment (oleg-jukovec) [inline net_test.go]: Please, don't test the private API, just a public API behavior.
```suggestion
package openssl_test
```

File: net_test.go [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+		ssl_listener.Addr().String(), ctx, InsecureSkipHostVerification)
+
+	wg.Wait()
+
+	if err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+	if client.is_shutdown {
+		t.Fatal("client is closed after creation")
+	}
```
Comment (oleg-jukovec) [inline net_test.go]: Let's better send a few bytes and read it from the echo server (It should work with sslConnect).

**Comment by oleg-jukovec** [comment]
Please, disable the windows runners (it is not our target platform or we need to fix tests on it).

File: net_test.go (hunk 21 above)
Comment (DerekBum) [reply net_test.go]: Updated.

File: net_test.go (hunk 20 above)
Comment (DerekBum) [reply net_test.go]: Updated.

File: net.go (hunk 19 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go (hunk 18 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go (hunk 17 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go (hunk 16 above)
Comment (DerekBum) [reply net.go]: Updated.

File: net.go [hunk 22]
```diff
@@ -89,8 +90,55 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
...
+}
+
+// DialContext acts like Dial but takes a context for network dial.
+//
+// The context includes only network dial. It does not include OpenSSL calls.
+//
+// See func Dial for a description of the network, addr, ctx and flags
+// parameters.
+func DialContext(context context.Context, network, addr string,
+	ctx *Ctx, flags DialFlags) (*Conn, error) {
```
Comment (oleg-jukovec) [inline net.go]: I would also like to suggest this renaming for the whole file:

1. There is now shadowing of the `context` module with the `context` variable.
2. On `Go` a variable `ctx` is usually `context.Context`. So the code would be clearer for a outside reader.
```suggestion
func DialContext(ctx context.Context, network, addr string,
	sslCtx *Ctx, flags DialFlags) (*Conn, error) {
```

**Review by oleg-jukovec — APPROVED** [review]
Thank you for the patch!

File: net.go (hunk 22 above)
Comment (DerekBum) [reply net.go]: Updated.

**Comment by DerekBum** [comment]
Disabled windows runners (in a separate commit).

**Review by DifferentialOrange — APPROVED** [review]

File: net.go [hunk 23]
```diff
@@ -15,22 +15,23 @@
...
+	"context"
 	"errors"
 	"net"
 	"time"
 )
 
 type listener struct {
 	net.Listener
-	ctx *Ctx
+	sslCtx *Ctx
```
Comment (oleg-jukovec) [inline net.go]: Let's revert this unnecessary change.

Comment (DerekBum) [reply net.go]: Updated.
